src/util/convertHTMLtoJSON.ts

```diff
@@ -0,0 +1,161 @@
+import _ from 'lodash'
+import { Element, HimalayaNode, Text, parse } from 'himalaya'
+import { Block } from '../action-creators/importText'
+
+interface PreBlock extends Block {
+  wasSpan?: boolean,
+  hasSpaces?: boolean,
+}
+
+/** Parses input HTML and saves in JSON array using Himalaya. */
+export const convertHTMLtoJSON = (html: string) => {
+
+  /** Check if string contains only characters from given 'includes' array. */
+  const isStringIncludesOnly = (str: string, includes: string[]) => {
+    const matched =
+      includes
+        .map(char => str.replace(new RegExp(`[^${char}]`, 'g'), '').length)
+        .reduce((acc, cur) => acc + cur, 0)
+    return matched === str.length
+  }
+
+  /** Retrieve attribute from Element node by key. */
+  const getAttribute = (key: string, node: Element) => {
+    const { attributes } = node
+    const attribute = attributes.find(attr => attr.key === key)
+    return attribute ? attribute.value : undefined
+  }
+
+  /** Removes empty nodes and comments from himalaya's JSON output. */
+  const removeEmptyNodesAndComments = (nodes: HimalayaNode[]) => {
+    return nodes.filter(node => {
+      if (node.type === 'element') {
+        if (node.tagName === 'br') return false
+        node.children = removeEmptyNodesAndComments(node.children)
+        return true
+      }
+      if (node.type === 'comment') return false
+      // remove text nodes containing only empty space and new line characters, such nodes could be created by himalaya in case of multi-line html input
+      if (node.type === 'text' && isStringIncludesOnly(node.content, ['\n', ' '])) return false
+      return node.content.length
+    }) as (Element | Text)[]
+  }
+
+  /** Append children to parent as children property if it's necessary. */
+  const joinChildren = (nodes: (PreBlock[] | PreBlock)[]): PreBlock[] | PreBlock => {
```

Comment:
Why does `joinChildren` take `PreBlock[] | PreBlock` instead of `PreBlock[]`? Same with the return value.

In general we want to avoid polymorphic behavior in function interfaces, i.e. accepting different types of inputs or returning different types of outputs. There are some exceptions in the codebase such as when functions handle either `Path` or `context`, but usually it is undesirable.

Reply:
I explain. E.g. such HTML:
```
<li>a<ul>
  <li>x</li>
  <li>y</li>
</ul></li>
```
converts by himalaya to such structure:
```
[
  {
    "type": "element",
    "tagName": "li",
    "attributes": [],
    "children": [
      {
        "type": "text",
        "content": "a"
      },
      {
        "type": "element",
        "tagName": "ul",
        "attributes": [],
        "children": [
          {
            "type": "element",
            "tagName": "li",
            "attributes": [],
            "children": [
              {
                "type": "text",
                "content": "x"
              }
            ]
          },
          {
            "type": "element",
            "tagName": "li",
            "attributes": [],
            "children": [
              {
                "type": "text",
                "content": "y"
              }
            ]
          }
        ]
      }
    ]
  }
]
```

In this JSON, thoughts **a** and **ul** are siblings. In `convert` function i iterate through tree and transform raw himalaya output to PreBlock.
For above output, after mapping and before `joinChildren` call, we have such data:
```
[
  {
    scope: 'a',
    children: null,
  },
  [
    {
      scope: 'x',
      children: null,
    },
    {
      scope: 'y',
      children: null
    }
  ]
]
```
So we have PreBlock (parent) and array of PreBlock (children). Array of PreBlock, which is placed next to PreBlock object in array, is determined as children of this object. 
Another example: 
```
[
  PreBlock (a), 
  [
    PreBlock (b), 
    [
      PreBlock, PreBlock, PreBlock
    ] (c), 
    PreBlock (d)
  ]
], 
```
results in
- a
  - b
    - PreBlock, PreBlock, PreBlock (c)
  - d


Two or more PreBlocks (**not** array) together are determined as siblings. 
E.g.
`[PreBlock, PreBlock, PreBlock, [Preblock, PreBlock]]` ->
```
- PreBlock
- PreBlock
- PreBlock
  - [PreBlock, PreBlock]
```

Reply:
Thank you for the good explanation. That helps me understand the himalaya output and the conversion to the nested PreBlock structure.

I have some reservations about the PreBlock structure. In the himalaya output, the node type makes the difference between siblings and children explicit. In the PreBlock structure, this becomes *implicit* and is determined indirectly by `Array.isArray`. The notion of "children" is defined in two ways now: either explicitly in the `children` property or in the array representation. This seems like a step backwards to me in terms of clarity. I will consider next steps and which one of us should undertake them, but feel free to let me if you have any thoughts.